### ISSUE TITLE
When running very few requests with a subscription onNext sometimes, requests have an overhead of about 1 second

### ISSUE THREAD

**Issue by OskarKjellin** [ReactiveX/RxNetty#387]
Creating a new ticket as the previous one was closed and there's no way for me to reopen it.

When we're testing out RxNetty we're getting some very unstable response times that sometimes reach a  factor of about 100 times slower than they should be. From what we can see from the logs, the request is not even sent until a couple of hundred milliseconds later. If we up the number of threads sending we don't see these peaks.

Small program to reproduce:

``` java
   HttpClient<ByteBuf, ByteBuf> rxNetty = RxNetty.createHttpClient("127.0.0.1", 1337);

        HttpServer<ByteBuf, ByteBuf> server = RxNetty.createHttpServer(1337, (HttpServerRequest<ByteBuf> request, HttpServerResponse<ByteBuf> response) -> {
            response.setStatus(HttpResponseStatus.OK);
            try
            {
                //Simulate network
                Thread.sleep(10);
            }
            catch (InterruptedException e)
            {
                e.printStackTrace();
            }
            response.writeString("OK");
            return response.close();
        });
        server.start();
        //Just set the logging to INFO to not log too much
        LoggerContext ctx = (LoggerContext) LogManager.getContext(false);
        Configuration config = ctx.getConfiguration();
        LoggerConfig loggerConfig = config.getLoggerConfig(LogManager.ROOT_LOGGER_NAME);
        loggerConfig.setLevel(Level.INFO);
        ctx.updateLoggers();

        List<Thread> threads = new ArrayList<>();
        int threadCount = 1;
        for (int i = 0; i < threadCount; i++)
        {
            Runnable runnable = new Runnable()
            {
                private final AtomicInteger requests = new AtomicInteger();

                @Override
                public void run()
                {
                    sendRequest();
                    while (requests.get() < 10000)
                    {
                        try
                        {
                            Thread.sleep(1000);
                        }
                        catch (InterruptedException e)
                        {
                            e.printStackTrace();
                        }
                    }
                }

                public void sendRequest()
                {
                    int req = requests.incrementAndGet();
                    if (req > 10000)
                    {
                        return;
                    }
                    long start = System.currentTimeMillis();
                    HttpClientRequest<ByteBuf> httpRequest = HttpClientRequest.createGet("/test");

                    rxNetty.submit(httpRequest)
                            .flatMap((HttpClientResponse<ByteBuf> r) ->
                                    r.getContent().map(b -> b.toString(StandardCharsets.UTF_8))
                            ).subscribe(c -> {
                        long dur = System.currentTimeMillis() - start;
                        if (dur > 100)
                        {
                            System.out.println(req + ":" + dur);
                        }
                        sendRequest();
                    }, ex -> ex.printStackTrace());


                }
            };
            Thread t = new Thread(runnable);
            t.start();
            threads.add(t);
        }
        for (Thread t : threads)
        {
            t.join();
        }
        server.shutdown();
```

Example output of printing all requests:

```
1023:12
1024:13
1025:13
1026:754
1027:11
1028:12
1029:11
...
5129:11
5130:11
5131:12
5132:11
5133:11
5134:1006
5135:804
5136:11
5137:11
5138:11
5139:11
5140:11
..
9726:11
9727:12
9728:1000
9729:803
9730:12
9731:10
9732:11
```

Adding timestamps to the peaks we clearly see a pattern of the millisecond of second.

```
2015-07-01T13:25:13.415+02:00 - 92:845
2015-07-01T13:25:14.416+02:00 - 128:557
2015-07-01T13:25:16.417+02:00 - 244:602
2015-07-01T13:25:17.249+02:00 - 251:758
2015-07-01T13:25:21.250+02:00 - 511:830
2015-07-01T13:25:22.253+02:00 - 513:989
2015-07-01T13:25:24.254+02:00 - 621:743
2015-07-01T13:25:25.255+02:00 - 673:402
2015-07-01T13:25:26.423+02:00 - 702:837
2015-07-01T13:25:27.256+02:00 - 703:833
2015-07-01T13:25:29.258+02:00 - 855:199
2015-07-01T13:25:30.258+02:00 - 905:433
2015-07-01T13:25:30.423+02:00 - 906:165
2015-07-01T13:25:31.259+02:00 - 919:695
2015-07-01T13:25:33.261+02:00 - 1025:787
2015-07-01T13:25:41.431+02:00 - 1654:859
2015-07-01T13:25:44.266+02:00 - 1865:441
2015-07-01T13:26:09.445+02:00 - 4068:356
2015-07-01T13:26:18.281+02:00 - 4773:893
2015-07-01T13:26:18.448+02:00 - 4774:167
2015-07-01T13:26:19.282+02:00 - 4817:355
2015-07-01T13:26:29.454+02:00 - 5630:983
2015-07-01T13:26:38.290+02:00 - 6377:411
2015-07-01T13:26:42.292+02:00 - 6655:832
2015-07-01T13:26:42.462+02:00 - 6656:170
2015-07-01T13:26:43.462+02:00 - 6658:986
2015-07-01T13:26:53.468+02:00 - 7528:255
```

Not subscribing to onNext but only to onCompleted does not show these peaks:
This has peaks:

``` java
                    RxNetty.createHttpGet("http://localhost:1337/test?id=" + req)
                            .flatMap((HttpClientResponse<ByteBuf> r) ->
                                    r.getContent().map(b -> b.toString(StandardCharsets.UTF_8))
                            ).doOnNext(c -> {
                        long dur = System.currentTimeMillis() - start;
                        if (dur > 100)
                        {
                            logger.warn(req + ":" + dur);
                        }
                        sum += dur;
                        sendRequest();
                    }).subscribe();
```

This runs as I expect

``` java
                    RxNetty.createHttpGet("http://localhost:1337/test?id=" + req)
                            .flatMap((HttpClientResponse<ByteBuf> r) ->
                                    r.getContent().map(b -> b.toString(StandardCharsets.UTF_8))
                            ).doOnCompleted(() -> {
                        long dur = System.currentTimeMillis() - start;
                        if (dur > 100)
                        {
                            logger.warn(req + ":" + dur);
                        }
                        sum += dur;
                        sendRequest();
                    }).subscribe();
```

Adding our own `MetricsEventListener` we see that these two events always occur next to eachother when we have a peak

```
2015-07-02 11:20:56,641 INFO  [rxnetty-nio-eventloop-3-4] client.LoadBalancingClientTests - FlushStart
2015-07-02 11:20:57,440 INFO  [rxnetty-nio-eventloop-3-2] client.LoadBalancingClientTests - BytesRead
```

**Comment by NiteshKant** [comment]
> > > Adding our own MetricsEventListener we see that these two events always occur next to eachother when we have a peak

@OskarKjellin Can you also record the events: FlushSuccess, FlushFailed, Write Start, WriteSuccess and Write Failed?

The `Thread.sleep` calls that you have in your code on the server side blocks the eventloop (shared b/w clients and server as you are in the same JVM), so I would not really trust the numbers that you are getting.

If you are looking at latency patterns, I would suggest to use something like [HDR histogram](https://github.com/HdrHistogram/HdrHistogram) instead to see the latency ranges. I don't think it's a good idea to look at outliers with such ad-hoc test programs. Most of the cases, there will be issues out of the scope of the test itself that conflate the numbers, eg: GC. I remember that you mentioned in the other issue that you are seeing this behavior in an actual server. In such a case, it will be good to first look at the server side metrics and look for GC events, network state, etc. whether the behavior was due to factors out of RxNetty.

**Comment by OskarKjellin** [comment]
@NiteshKant I will give a more thorough answer when I'm back from my vacation next week and have my computer.  

I did record all events but the only pattern I could see was that it was always those two after one another when there was a peak. 

I also realized that sleeping the thread might be bad for the thread pool so I changed it to the equivalent observable. Don't remember the name of it but I could see the same issue anyhow. Also there are more threads than concurrent requests so it shouldn't matter? 

I will try with HDR histograms. I first noticed this issue with the hystrix dashboard (I am running hystrix and ribbon in our real environment). Also added the gc logs but they were at completely different times from the peaks. 

I understand that of course there might be something wrong with the test itself rather than the library but because we're seeing the exact same issue in our real environment I am quite certain the test is correct. I tried to reproduce it without all the other factors that come into play when running a full blown environment and I don't get these peaks if I remove rxnetty. 

Even though all these other factors, would they really make rxnetty log these two events with a high interval in between and at the same millisecond if second? If that was the gc it would not only occur at the same millis but also when rxnetty was doing the exact same thing which seems very unlikely. It also seems very unlikely that these issues would only occur when using onNext and not when using onComplete? Given the characteristics I find in my small tests it seems unlikely that the test is wrong but I might just as well be using rxnetty wrong.

**Comment by NiteshKant** [comment]
> I will give a more thorough answer when I'm back from my vacation next week and have my computer.

Great, I will wait for your detailed answer!

> > Also there are more threads than concurrent requests so it shouldn't matter?

In netty, a channel is bound to an eventloop for its lifetime, so if you block that eventloop (thread) then it does not matter how many other threads are there in the system, nothing will happen on that channel/connection till the time the eventloop is unblocked. The latency you see is an obvious outcome of this blocking.

> > Even though all these other factors, would they really make rxnetty log these two events with a high interval in between and at the same millisecond if second?

The event flush start happens on the user thread, flush completes happen on the eventloop. Assuming request-response scenario, bytes read will only happen after flush is done on the request. If flush started but did not complete, it may be because of eventloop queue being backed up, which may be because the eventloop is blocked. So, don't block the eventloop :)

My point here is that do not co-relate the results or hypothesis from the test to determine what happened at the server, because there are obvious issues with the test. If you have the metrics from the server, eg: pending writes, pending flushes, etc. at the time of the latency spike, it will be more useful than the test.

**Comment by OskarKjellin** [comment]
@NiteshKant now I'm back :)

I changed my `sleep` to:

```
             return Observable.timer(10, TimeUnit.MILLISECONDS).flatMap(c -> response.close());
```

Not sure if that blocks the eventloop or not.

After reading your explanation I've been wondering a bit if the same thing is actually happening on the server. Let me give a bit of more background:
I'm running an API layer as a proxy to internal components. On the API I'm running jax rs with jersey and sending the requests with RXNetty. I then subscribe to `onNext` where I set the response of `AsyncResponse`. This in turn triggers all the `MessageBodyWriters` and so forth and then ultimately it writes the response to the `OutputStream`. This all feels way too much to do in the eventloop thread or am I mistaken here? What is the intended way to do this? I currently changed it to `observeOn(Schedulers.io())` and have yet to see a spike since.

**Comment by NiteshKant** [comment]
> > This all feels way too much to do in the eventloop thread or am I mistaken here?

The problem here would be that you are writing to the socket using blocking I/O inline while processing the request. So, if the write blocks, the eventloop is blocked. Using `observeOn`/`subscribeOn` is the correct way of handling a blocking call during emission and subscription respectively.

If you are satisfied that blocking the eventloop was indeed the issue, do you mind if this issue is closed?

**Comment by OskarKjellin** [comment]
Yeah, that's what I figured as well. Do I need to use `subscribeOn` as well?

Go ahead and close it. Thanks a lot for your help, sorry for spamming you :)

**Comment by NiteshKant** [comment]
If you are potentially blocking in the `subscribe` call use `subscribeOn`, if you are blocking on an emission (`onNext`, `onComplete`, `onError`) then use `observeOn`.
